From 8c848ee6580f3a397008d3dcbd6fa92d2603f909 Mon Sep 17 00:00:00 2001
From: Brandon Black <bblack@wikimedia.org>
Date: Fri, 12 Sep 2025 07:19:01 -0500
Subject: [PATCH 103/133] std.os.linux.setsid(): return raw syscall0 result
X-Developer-Signature: v=1; a=openpgp-sha256; l=2057; i=xtex@astrafall.org;
 h=from:subject; bh=8+GLNZotuV0AcfMHcHqrUsSXL2TZ7TGYmOp9iltUqLQ=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78rhnFt/X0/t40blcm72pRlFgfcLWLy7zVJQWM7XKM
 /yvsI3smMjCIMbFYCmmyFJk2ODNqpPOL7qsXBZmDisTyBBpkQYGBgYGFga+3MS8UiMdIz1TbUM9
 Q0MdIx0jBi5OAZjqlzsYGRamPttya3n3kV+F+z9lCGQk5hk23Mv7ezv0UJwcp98e+fsM/3M2pE6
 ctfzTtYamVYIpz/UOFPX8qphpm3CB0c7+s9bdxdwA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

When not linking libc on 64-bit Linux and calling posix.setsid(),
we get a type error at compile time inside of posix.errno().  This
is because posix.errno()'s non-libc branch expects a usize-sized
value, which is what all the error-returning os.linux syscalls
return, and linux.setsid() instead returned a pid_t, which is only
32 bits wide.

This and the other 3 pid-related calls just below it (getpid(),
getppid(), and gettid()) are the only Linux syscall examples here
that are casting their return values to pid_t. For the other 3
this makes sense: those calls are documented to have no possible
errors and always return a valid pid_t value.

However, setsid() actually can return the error EPERM, and
therefore needs to return the raw value from syscall0 for
posix.errno() to process like normal.

Additionally, posix.setsid() needs an @intCast(rc) for the success
case as a result, like most other such cases.

(cherry picked from commit 04071d64bb83c0600d445a2f4b2541e68a6021bf)
---
 lib/std/os/linux.zig | 4 ++--
 lib/std/posix.zig    | 2 +-
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/lib/std/os/linux.zig b/lib/std/os/linux.zig
index 1b0b752f76fd..a150ad2782a3 100644
--- a/lib/std/os/linux.zig
+++ b/lib/std/os/linux.zig
@@ -1834,8 +1834,8 @@ pub fn setgroups(size: usize, list: [*]const gid_t) usize {
     }
 }
 
-pub fn setsid() pid_t {
-    return @bitCast(@as(u32, @truncate(syscall0(.setsid))));
+pub fn setsid() usize {
+    return syscall0(.setsid);
 }
 
 pub fn getpid() pid_t {
diff --git a/lib/std/posix.zig b/lib/std/posix.zig
index 818065fb9a54..a5c09c5cf2b9 100644
--- a/lib/std/posix.zig
+++ b/lib/std/posix.zig
@@ -6833,7 +6833,7 @@ pub const SetSidError = error{
 pub fn setsid() SetSidError!pid_t {
     const rc = system.setsid();
     switch (errno(rc)) {
-        .SUCCESS => return rc,
+        .SUCCESS => return @intCast(rc),
         .PERM => return error.PermissionDenied,
         else => |err| return unexpectedErrno(err),
     }
-- 
2.52.0

